### PR TITLE
Take a deposit when creating a pool

### PR DESCRIPTION
**Pull request by branan**
Previously, pool creation was free. This allows for spamming storage
with a large number of pools. To prevent this, we now take a deposit
when creating a pool.

To allow for the pool pallet itself to not care whether it is created
by a signed origin or a special origin (such as democracy), we first
attempt to take the deposit from the origin if possible. If the origin
cannot provide a deposit, we instead take it from the specified admin
account.

### CODE REVIEW

**Comment by branan** [comment]
Unfortunately there are dueling changes from the refactor branch and the parachain branch, so there's not really a clean base - whichever place I target this, it will need a rebase when parachain+rebase are merged.

I think at this point the best thing is to just wait until the refactor branch is merged. I can then rebase this on top of the final state of things. It's only going to be a few lines of small conflicts.

**Review by NunoAlexandre — COMMENTED** [review]
Some thoughts:

0. Do we want to be able to unlock a deposit once a pool is removed?

1. With the current approach of having a total sum of all deposits per account, it makes it impossible to implement point `0.` in the future, shall we ever want to be free to tweak the deposit value. We can future-proof this solution by implementing `Deposit` as a `DoubleMap<Account, PoolId, Amount>`?

2. It would be good to extend the tests to verify that the deposit amount is locked up after the pool is created.

**Review by mustermeiszer — COMMENTED** [review]
Did start a discussion whether we really want the depositor etc. appended to the `PoolDetails`. 

While thinking about this I was wondering if we want to consider the following approach. 
* Adapt pallet-fees to be a universal fees/deposit location
* Introduce a trait like `CostCollector`
* call this with the amount we want to be deposited

```rust
trait CostCollector<AccountId> {
      /// The balance type
      type Balance;
      
      /// The identifier for a pledge.
      type Id;
      
      /// Takes amount of fees from who. What we with those is up to us
      fn fee(who: AccountId, amount: Self::Balance) -> DispatchResult;
      
      /// Take amount as pledge,
      fn deposit_pledge(who: AccountId, amount: Self::Balance, id: Self::Id) -> DispatchResult;
      
      /// Release the pledge
      fn release_pledge(id: Self::Id) -> DispatchResult;
}
```

The id in our case could be `HashOf(Depositer & pool-id)`. This way we only must store the depositer in the pool in a `StorageMap(T::PoolId, T::AcccountId)` and could hash this to get the id back and release the pledge. 

Also, in the future we could use the fees, for a general purpose fees/pledge location maybe even extending it to be used with the upcoming fees infrastructure that we envision as network fees. 

Anyways, just got my ideas out. Tell me what you think.

File: pallets/pools/src/lib.rs
Comment (offerijns) [reply pallets/pools/src/lib.rs:581]: Makes sense!

**Comment by branan** [comment]
@mustermeiszer Right now I think the deposit-taking logic is simple enough that abstracting it away doesn't make a ton of sense. I'd say making a pallet probably isn't worth it until we have to do something weird like track deposits from across multiple pallets?

**Comment by mustermeiszer** [comment]
I would use the existing fees pallet we have. But I see your point. 

My main concern is that once we have this storage, transitioning to another storage will be painful, so we will stick with this. 

As adding and removing storage is not trivial, if I remember correctly. 

But not a blocker from my side.

**Comment by branan** [comment]
I want to add a quick test around the deposit behavior before this lands, but otherwise it's rebased and is complete

**Review by offerijns — APPROVED** [review]
LGTM!

**Review by mustermeiszer — APPROVED** [review]
LGTM!

**Pull request merged**
